keps/sig-network/20191205-kube-proxy-IP-node-binding.md

```diff
@@ -0,0 +1,70 @@
+---
+title: Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+authors:
+  - "@Sh4d1"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloudprovider
+reviewers:
+  - "@thockin"
+  - "@cheftako"
+  - TBD
+approvers:
+  - "@thockin"
+  - TBD
+editor: TBD
+creation-date: 2019-12-05
+last-updated: 2019-12-05
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+The different kube proxy implentations (at least ipvs and iptables) are today sort of binding the External IPs of the LoadBalancer Service to each node. (iptables is creating some rules to redirect directly the packets to the service and ipvs is binding the IP to one interface on the node). This feature is exisiting because:
+
+1. To redirect pod traffic that is intended for LoadBalancer IP directly to the service, bypassing the loadbalancer for path optimization.
+2. Some LBs are sending packet with the destination IP equal to the LB IP, so the IP needs to be routable directly to the right service (otherwise looping back and forth)
+3. (TBD)
+
+This enhancement proposes a way to make this features configurable, giving the cloud provider, or the user, a way to disable this behaviour of kube proxy.
```

Comment:
Does this imply that we want this behavior for all services (essentially environment / provider choice) or do we want this to be per service behavior? 

From the use cases, it sounds like we want this for all services in a cluster?

Reply:
A good solution I think would be to default all LB service to a value given by the cloud controller manager (and if not given, the current behaviour) and allow the user the overwrite this behaviour on a per service basis

Reply:
actually I was thinking of it slightly differently. If environment specific, then it's a flag in kube-proxy (applies to all services that kube-proxy tracks). Or if it is per specific, then it could be something like an annotation on the service (that kube-proxy reads and writes different rules for that service). Given that we don't have any interactions between kube-proxy and cloud-controller-manager (AFAIK), how would it be something that is implemented in cloud controller manager?

Reply:
There is no interaction between kube proxy and ccm, but the cloud controller could call a specific method in the CCM and set the value on the service (field or annotation) that could then be read by kube proxy